integration/e2e/metrics_promql.go

```diff
@@ -0,0 +1,283 @@
+					val:    float64(m.GetSummary().GetSampleCount()),
+				})
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_sum", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetSummary().GetSampleSum(),
+				})
+			}
+
+		case dto.MetricType_HISTOGRAM:
+			for _, m := range mf.GetMetric() {
+				for _, b := range m.GetHistogram().GetBucket() {
+					result = append(result, sample{
+						labels: getLabelsExt(mf.GetName()+"_bucket", m.GetLabel(), true, b.GetUpperBound()),
+						ts:     getTimestamp(ts, m.GetTimestampMs()),
+						val:    float64(b.GetCumulativeCount()),
+					})
+				}
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_count", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    float64(m.GetHistogram().GetSampleCount()),
+				})
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_sum", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetHistogram().GetSampleSum(),
+				})
+			}
+		}
+	}
+
+	return result
+}
+
+func getTimestamp(ts int64, mts int64) int64 {
+	if mts != 0 {
+		return mts
+	}
+	return ts
+}
+
+func getLabels(name string, pairs []*dto.LabelPair) map[string]string {
+	return getLabelsExt(name, pairs, false, 0)
+}
+
+func getLabelsExt(name string, pairs []*dto.LabelPair, bucket bool, upperBound float64) map[string]string {
+	result := map[string]string{}
+	result[labels.MetricName] = name
+	if bucket {
+		result[labels.BucketLabel] = fmt.Sprintf("%g", upperBound)
+	}
+
+	for _, p := range pairs {
+		result[p.GetName()] = p.GetValue()
+	}
+	return result
+}
+
+// implements Queryable and Querier
+type samplesQueryable []sample
+
+func (sq *samplesQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return sq, nil
+}
+
+func (sq samplesQueryable) SelectSorted(params *storage.SelectParams, matcher ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
+	var series []storage.Series
+
+	for _, s := range sq {
+		if s.matches(matcher...) {
+			series = append(series, s)
+		}
+	}
+
+	return &seriesSet{series: series}, nil, nil
+}
+
+func (sq samplesQueryable) Select(params *storage.SelectParams, matcher ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
+	return sq.SelectSorted(params, matcher...)
+}
+
+func (sq *samplesQueryable) LabelValues(name string) ([]string, storage.Warnings, error) {
+	panic("implement me")
+}
+
+func (sq *samplesQueryable) LabelNames() ([]string, storage.Warnings, error) {
+	panic("implement me")
+}
+
+func (sq *samplesQueryable) Close() error {
+	return nil
+}
+
+// seriesSet implements storage.SeriesSet.
+type seriesSet struct {
+	cur    int
+	series []storage.Series
+}
+
+func (c *seriesSet) Next() bool {
+	c.cur++
+	return c.cur-1 < len(c.series)
+}
+
+func (c *seriesSet) At() storage.Series {
+	return c.series[c.cur-1]
```

Comment:
You could remove the `-1` which is tricky to read just initializing the `seriesSet{cur : -1}`

Reply:
I copied this code from somewhere, without much thinking. Turns out, I actually forgot to sort series!

We could also use `github.com/cortexproject/cortex/pkg/querier/series.NewConcreteSeriesSet` method, but that introduces dependency to cortex package.